darshan-util/pydarshan/darshan/backend/cffi_backend.py

```diff
@@ -650,3 +650,70 @@ def _log_get_heatmap_record(log):
+
+
+def log_get_derived_metrics(log_path: str, mod_name: str):
+    """
+    Returns the darshan_derived_metrics struct from CFFI/C accumulator code.
+
+    Parameters:
+        log_path: Path to the darshan log file
+        mod_name: The name of the module to retrieve derived metrics for
+
+    Returns:
+        darshan_derived_metrics struct (cdata object)
+    """
+    # TODO: eventually add support for i.e., a regex filter on the records
+    # the user wants to get derived metrics for--like filtering to records
+    # with a single filename involved before accumulating the data?
+    log_handle = log_open(log_path)
+    jobrec = ffi.new("struct darshan_job *")
+    libdutil.darshan_log_get_job(log_handle['handle'], jobrec)
+    modules = log_get_modules(log_handle)
+
+    if mod_name not in modules:
+        return None
+    mod_type = _structdefs[mod_name]
+
+    buf = ffi.new("void **")
+    r = libdutil.darshan_log_get_record(log_handle['handle'], modules[mod_name]['idx'], buf)
+    rbuf = ffi.cast(mod_type, buf)
+
+    darshan_accumulator = ffi.new("darshan_accumulator *")
+    print("before create")
+    r = libdutil.darshan_accumulator_create(modules[mod_name]['idx'],
+                                            jobrec[0].nprocs,
+                                            darshan_accumulator)
+    if r != 0:
+        raise RuntimeError("A nonzero exit code was received from "
+                           "darshan_accumulator_create() at the C level. "
+                           f"This could mean that the {mod_name} module does not "
+                           "support derived metric calculation, or that "
+                           "another kind of error occurred. It may be possible "
+                           "to retrieve additional information from the stderr "
+                           "stream.")
+    print("after create")
+
+    print("before inject")
+    r = libdutil.darshan_accumulator_inject(darshan_accumulator[0], rbuf[0], 1)
+    if r != 0:
+        raise RuntimeError("A nonzero exit code was received from "
+                           "darshan_accumulator_inject() at the C level. "
+                           "It may be possible "
+                           "to retrieve additional information from the stderr "
+                           "stream.")
+    print("after inject")
+    darshan_derived_metrics = ffi.new("struct darshan_derived_metrics *")
+    print("before emit")
+    r = libdutil.darshan_accumulator_emit(darshan_accumulator[0],
+                                          darshan_derived_metrics,
+                                          rbuf[0])
+    if r != 0:
+        raise RuntimeError("A nonzero exit code was received from "
+                           "darshan_accumulator_emit() at the C level. "
+                           "It may be possible "
+                           "to retrieve additional information from the stderr "
+                           "stream.")
+    print("after emit")
```

Comment:
After this prints in the test for `POSIX` module I see `free(): corrupted unsorted chunks` and then a segfault. I'm not sure `rbuf[0]` is the right argument for `darshan_accumulator_emit` third parameter, but either way the exit code is `0`.

Since emit runs with a `0` exit code, perhaps CFFI tries to `free` one of the weird exposed data structures that gets passed around here, not sure.

Reply:
If I use the custom darshan desctrutor instead I get `malloc(): invalid size (unsorted)`.

Reply:
My guess would have been that the `rbuf[0]` arguments would either be `buf[0]` or `rbuf`.  I'm still trying to figure out the CFFI conventions, but at any rate you want a *pointer* to a single record buffer going into the C calls.  I would think that `rbuf[0]` as is used now in that PR is a deferenced pointer (i.e. the value of the struct rather than a pointer to the struct), but I'm kind of guessing.

@tylerjereddy when you use the custom darshan destructor, you mean you added a call to `libdutil.darshan_free(buf[0])`?  That much seems like it should be fine based on earlier examples in that file.

The only other thing I notice is that the return code for `get_record()` is not checked, but that's probably unrelated to your immediate problem; the `inject` call would have failed if the record didn't match up with the accumulator.